### ISSUE TITLE
Mapping nested JSON objects

### ISSUE THREAD

**Issue by randre03** [aconchillo/guile-json#80]
I access an api which sends its response back in the following shape:
```
{
  "results": {
    "sunrise": "9:52:03 AM",
    "sunset": "12:20:22 AM",
    "first_light": "8:01:33 AM",
    "last_light": "2:10:52 AM",
    "dawn": "9:21:29 AM",
    "dusk": "12:50:56 AM",
    "solar_noon": "5:06:12 PM",
    "golden_hour": "11:41:50 PM",
    "day_length": "14:28:19",
    "timezone": "UTC"
  },
  "status": "OK"
}
```
I would like this data (or, in the future, any nested json) to be accessible using guile's record type. When I feed this into my record constructor, I wind up with the 'results' field containing all the data and fields like 'sunrise' and 'dawn' are all unspecified. Is there a way to go from the JSON above directly to the record type without first extracting all the fields under the "results" section and using that as the data provided to the record constructor?

Thanks for a great module!

**Comment by aconchillo** [comment]
Hi @randre03 ! Thank you for reporting this. Can you please share your record definition?

**Comment by randre03** [comment]
```
(define-json-type <day>
  (results)
  (status)
  (sunrise)
  (sunset)
  (first-light)
  (last-light)
  (dawn)
  (dusk)
  (solar-noon)
  (golden-hour)
  (day-length)
  (timezone))
```

**Comment by aconchillo** [comment]
One way to do this would be:

```
(define-json-type <results>
  (sunrise)
  (sunset)
  ...)

(define-json-type <day>
  (results "results" <results>)
  (status)
  (sunrise)
  (sunset)
  (first-light)
  (last-light)
  (dawn)
  (dusk)
  (solar-noon)
  (golden-hour)
  (day-length)
  (timezone))
```

Now the `results` filed would be a record type with the fields you defined in `<record>` (and you only need to specify the ones you are interested in). Not sure if that's what you wanted.

Is it possible that `results` is different every time? If so, there's another way you could do it.

**Comment by randre03** [comment]
Thanks, I had considered that approach but wondered if there is a more "direct" method requiring only one record type. What is the other way you allude to as there are other apis I need to work where the shape of the responses are the same but the values change (for example, an api that reports the current price of stocks, currency, etc).

**Comment by aconchillo** [comment]
Oh... I just tried it and I see now. You should have:

```
(define-json-type <day>
  (results "results" <day-results>)
  (status))
```

and then:

```
(define-json-type <day-results>
  (sunrise)
  (sunset)
  (first-light)
  (last-light)
  (dawn)
  (dusk)
  (solar-noon)
  (golden-hour)
  (day-length)
  (timezone))
```

Your initial record doesn't actually match the JSON example you provided. Tthe fields are not nested, that's why they show up as `*unspecified*` because they are not at the same level as `results` they are one level below.

That would definitely be the best way to do it with records, but you could always go the `define-json-mapping` way where you have much more control.

See how I named it `day-results`. If you have different `results` you should then have different `day` records and different `results` records.

**Comment by aconchillo** [comment]
And that made me realize there's actually a bug if you use `(json->day port)`. I'll try to fix it later, but in any case I believe what I proposed above would be the best way to do it.

**Comment by randre03** [comment]
Gotcha. I've got things working with the example you provided. I think I'll play around with `define-json-mapping` as well to see how far I can get. Thanks for your help and quick responses.

**Comment by aconchillo** [comment]
Sounds good. Btw, I just released 4.7.3 that fixes a regression introduced in 4.7.1. I'm not sure what version you are using.
